modules/sdk-lib-mpc/src/tss/ecdsa/noSmallFactorsProof.ts

```diff
@@ -0,0 +1,145 @@
+/**
+ * Implementation of No Small Factors ($\Pi^\text{fac}).
+ * https://eprint.iacr.org/2020/492.pdf Section B.4
+ */
+import { createHash } from 'crypto';
+import { bitLength, randBetween } from 'bigint-crypto-utils';
+import { modPow } from 'bigint-mod-arith';
+import { bigIntFromBufferBE, bigIntToBufferBE } from '../../util';
+import { DeserializedNoSmallFactorsProof } from './types';
+
+const ORDER = BigInt('0xfffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141');
+const ELL = BigInt(256);
+const EPSILON = BigInt(BigInt(2) * ELL);
+
+/**
+ * Generate pseudo-random challenge value $e$ for $(N, w)$.
+ * @param N - the prime number to verify is a product of two large primes.
+ * @param w - a random number with the same bitLength as N, that satisfies the Jacobi of w is -1 wrt N.
+ * @returns {bigint} - challenge value $e$.
+ */
+function generateE(N: bigint, w: bigint): bigint {
+  const digest = createHash('shake256', { outputLength: 1 + Math.floor((bitLength(ORDER) + 7) / 8) })
+    .update(bigIntToBufferBE(N))
+    .update('$')
+    .update(bigIntToBufferBE(w))
+    .update('$')
+    .digest();
+  const e = bigIntFromBufferBE(digest.subarray(1)) % ORDER;
+  if (digest[0] & 1) {
+    return -e;
+  }
+  return e;
+}
+
+/**
+ * Calculate the closest integer square root of $n$.
+ * @param n - the number to calculate the square root of.
+ * @returns {bigint} - $n$'s closest integer square root.
+ */
+function isqrt(n: bigint): bigint {
+  if (n < BigInt(0)) {
+    throw new Error();
+  }
+  if (n < BigInt(2)) {
+    return n;
+  }
+  function newtonIteration(n: bigint, x0: bigint) {
+    const x1 = (n / x0 + x0) >> BigInt(1);
+    if (x0 === x1 || x0 === x1 - BigInt(1)) {
+      return x0;
+    }
+    return newtonIteration(n, x1);
+  }
+  return newtonIteration(n, BigInt(1));
+}
+
+/**
+ * Prove that $n0$ has no small factors, where $n0$ is the product of two large primes.
+ * @param p - a large prime.
+ * @param q - a large prime.
+ * @param w - a random number with the same bitLength as $p * q$, that satisfies the Jacobi of w is -1 wrt $n0$.
+ * @param nHat - a safe bi-prime, such as that returned from rangeProof.generateNTilde.
+ * @param s - security parameters for $nHat$ such as the $h1$ value returned from rangeProof.generateNTilde.
+ * @param t - security parameters for $nHat$ such as the $h2$ value returned from rangeProof.generateNTilde.
+ * @returns proof that the product of $p * q$ has no small factors.
+ */
+export function prove(
+  p: bigint,
+  q: bigint,
+  w: bigint,
+  nHat: bigint,
+  s: bigint,
+  t: bigint
+): DeserializedNoSmallFactorsProof {
+  const n0 = p * q;
+  const e = generateE(n0, w);
+  const sqrtN0 = isqrt(n0);
+  const alpha = randBetween(sqrtN0 << (ELL + EPSILON), -sqrtN0 << (ELL + EPSILON));
+  const beta = randBetween(sqrtN0 << (ELL + EPSILON), -sqrtN0 << (ELL + EPSILON));
+  const sigma = randBetween((nHat * n0) << ELL, -(nHat * n0) << ELL);
```

Comment:
It's better to name it `rho` to be consistent with the paper.

Reply:
Done